File: Recursive Algorithms/minimax/tic-tac-toe/tictactoe.py
import time
import random


class TicTacToe():
    '''Class that has all necesarry components for a tic tac toe game'''
    def __init__(self, player1, player2):
        self.board = [[None, None, None],
                      [None, None, None],
                      [None, None, None]]
        self.current_player = player1.player_mark
        self.player1 = player1
        self.player2 = player2

    def render(self):
        '''Renders the current board state'''
        for row in self.board:
            print("|", end="")
            for item in row:
                if item is None:
                    print(".", end="|")
                else:
                    print(item, end="|")
            print()

    def possible_moves(self):
        '''Finds all possible_moves of current board state'''
        moves = []
        for x in range(len(self.board)):
            for y in range(len(self.board[x])):
                if self.board[x][y] is None:
                    moves.append((x, y))

        return moves

    def legal_move(self, move):
        '''Determines if given move is legal'''
        legal_moves = self.possible_moves()
        if move in legal_moves:
            return True
        return False

    def move(self, move):
        '''Moves a piece and switches current_player'''
        if self.legal_move(move):
            move_x = move[0]
            move_y = move[1]
            self.board[move_x][move_y] = self.current_player
            self.last_move = move
            self.switch()
            return True
        else:
            return False

    def switch(self):
        '''Switches current player'''
        if self.current_player == "x":
            self.current_player = "o"
        else:
            self.current_player = "x"

    def game_over(self):
        '''Finds if current board_state is full and the game is over'''
        if self.game_won():
            return True
        for row in self.board:
            for item in row:
                if item is None:
                    return False
        return True

    def game_won(self):
        '''Finds if game is won and updates self.winning_player'''
        board = self.board
        for x in range(len(self.board)):
            for y in range(len(self.board[x])):
                mark = self.board[x][y]
                if board[x][y] is not None:
                    if board[0][y] == board[1][y] == board[2][y]:
                        self.winning_player = mark
                        return True
                    if board[x][0] == board[x][1] == board[x][2]:
                        self.winning_player = mark
                        return True
                    if x == y and board[0][0] == board[1][1] == board[2][2]:
                        self.winning_player = mark
                        return True
                    if x + y == 2:
                        if board[0][2] == board[1][1] == board[2][0]:
                            self.winning_player = mark
                            return True
        return False

    def revert_move(self, move):
        x = move[0]
        y = move[1]
        self.switch()
        self.board[x][y] = None

    def play(self):
                
        while not self.game_over():
            player_1_move = self.player1.best_move(self)
            time.sleep(1)
            self.move(player_1_move)
            self.render()
            print()
            
            if self.game_over():
                if self.game_won():
                    print(self.winning_player, "has won!")
                    exit(0)
                print("Nobody has won")
                exit(0)                

            player_2_move = self.player2.best_move(self)
            time.sleep(1)
            self.move(player_2_move)
            self.render()
            print()

        if self.game_won():
            print(self.winning_player, "has won!")
            exit(0)
        else:
            print("Nobody has won.")


class MiniMaxPlayer():
    '''A perfect player that uses minimax to determine best move'''
    def __init__(self, player):
        self.player_mark = player

    def score(self, game, depth):
        '''returns score of given board'''
        if game.game_won():
            if game.winning_player is self.player_mark:
                return 10 - depth
            else:
                return depth - 10
        return 0

    def minimax(self, game, depth):
        '''Minimax solution to selecting what to play'''
        if game.game_over():
            return self.score(game, depth)

        depth += 1

        scores = []
        moves = []
        # Recursively adds all scores of every possible move of a given state
        for move in game.possible_moves():
            game.move(move)
            scores.append(self.minimax(game, depth))
            game.revert_move(move)
            moves.append(move)

        self.scores = scores
        self.moves = moves

        # max solution in case current player is this player
        if game.current_player is self.player_mark:
            max_score_idx = scores.index(max(scores))
            highest_score = scores[max_score_idx]
            # finds all moves with same value and selects one at random so
            # every game doesn't play out the same
            best_moves = []
            for i, move in enumerate(moves):
                if scores[i] == highest_score:
                    best_moves.append(move)
            self.choice = (random.choice(best_moves))
            return scores[max_score_idx]

        # min solution in case current player is opponent
        else:
            min_score_idx = scores.index(min(scores))
            self.choice = moves[min_score_idx]
            return scores[min_score_idx]

    def best_move(self, game):
        '''returns best move calculated with mini-max'''
        self.minimax(game, 0)
        return self.choice


class HumanPlayer():
    def __init__(self, player):
        self.player_mark = player

    def best_move(self, game):
        possible_moves = game.possible_moves()
        x = None
        y = None
        while True:
            x = int(input("What row do you want to mark?: ")) - 1
            y = int(input("What column do you want to mark?: ")) - 1
            if (x, y) in possible_moves:
                break
            else:
                print("That position has already been marked")
        return (x, y)


# player1 starts, indepent of marks.
player1 = HumanPlayer("x")
player2 = MiniMaxPlayer("o")

game = TicTacToe(player1, player2)
game.play()

File: Combinatorial Game Theory/Tic-Tac-Toe Optimal Move/Cpp/Tic-Tac-Toe.cpp
// C++ program to find the next optimal move for a player
#include <bits/stdc++.h>
using namespace std;

struct Move
{
  int row, col;
};

char player = 'x', opponent = 'o';

bool isMovesLeft(char board[3][3])
{
  for (int i = 0; i < 3; i++)
    for (int j = 0; j < 3; j++)
      if (board[i][j] == '_')
        return true;
  return false;
}

int evaluate(char b[3][3])
{
  for (int row = 0; row < 3; row++)
  {
    if (b[row][0] == b[row][1] &&
        b[row][1] == b[row][2])
    {
      if (b[row][0] == player)
        return +10;
      else if (b[row][0] == opponent)
        return -10;
    }
  }

  for (int col = 0; col < 3; col++)
  {
    if (b[0][col] == b[1][col] &&
        b[1][col] == b[2][col])
    {
      if (b[0][col] == player)
        return +10;

      else if (b[0][col] == opponent)
        return -10;
    }
  }

  if (b[0][0] == b[1][1] && b[1][1] == b[2][2])
  {
    if (b[0][0] == player)
      return +10;
    else if (b[0][0] == opponent)
      return -10;
  }

  if (b[0][2] == b[1][1] && b[1][1] == b[2][0])
  {
    if (b[0][2] == player)
      return +10;
    else if (b[0][2] == opponent)
      return -10;
  }
  return 0;
}

int minimax(char board[3][3], int depth, bool isMax)
{
  int score = evaluate(board);

  if (score == 10)
    return score;

  if (score == -10)
    return score;

  if (isMovesLeft(board) == false)
    return 0;

  if (isMax)
  {
    int best = -1000;
    for (int i = 0; i < 3; i++)
    {
      for (int j = 0; j < 3; j++)
      {
        if (board[i][j] == '_')
        {
          board[i][j] = player;

          best = max(best, minimax(board, depth + 1, !isMax));
          board[i][j] = '_';
        }
      }
    }
    return best;
  }

  else
  {
    int best = 1000;

    for (int i = 0; i < 3; i++)
    {
      for (int j = 0; j < 3; j++)
      {
        if (board[i][j] == '_')
        {
          board[i][j] = opponent;
          best = min(best, minimax(board, depth + 1, !isMax));

          board[i][j] = '_';
        }
      }
    }
    return best;
  }
}

Move findBestMove(char board[3][3])
{
  int bestVal = -1000;
  Move bestMove;
  bestMove.row = -1;
  bestMove.col = -1;
  for (int i = 0; i < 3; i++)
  {
    for (int j = 0; j < 3; j++)
    {
      if (board[i][j] == '_')
      {
        board[i][j] = player;
        int moveVal = minimax(board, 0, false);
        board[i][j] = '_';
        if (moveVal > bestVal)
        {
          bestMove.row = i;
          bestMove.col = j;
          bestVal = moveVal;
        }
      }
    }
  }

  printf("The value of the best Move is : %d\n\n", bestVal);
  return bestMove;
}
int main()
{
  char board[3][3] =
      {
          {'x', 'o', 'x'},
          {'o', 'o', 'x'},
          {'_', '_', '_'}
      };

  Move bestMove = findBestMove(board);

  printf("The Optimal Move is :\n");
  printf("ROW: %d COL: %d\n\n", bestMove.row, bestMove.col);
  return 0;
}
